src/ObjectUtils/CoffFile.cpp

```diff
@@ -417,49 +417,130 @@ orbit_object_utils::CoffFileImpl::GetPrimaryRuntimeFunction(
+  // merge two RUNTIME_FUNCTIONs if:
+  // - they are adjacent (the end address of the first equals the start address of the second); and
+  // - they have the same primary RUNTIME_FUNCTION (where a RUNTIME_FUNCTION that doesn't have
+  //   chained UNWIND_INFO is its own primary), as this corresponds to the prologue of the function.
+  //
+  // Note that this rule doesn't allow for "holes": if two consecutive RUNTIME_FUNCTIONs are not
+  // perfectly adjacent, we will consider them two separate functions, even if they have the same
+  // primary RUNTIME_FUNCTION. However, we never observed this case.
+  //
+  // And note that we don't require the primary RUNTIME_FUNCTION to actually be part of the current
+  // set of RUNTIME_FUNCTIONs we are merging. We observed this as a very rare but valid case, on
+  // which more information can be found below.
   std::vector<UnwindRange> unwind_ranges;
-  // For a function, the "primary" RUNTIME_FUNCTION and the ones with UNW_FLAG_CHAININFO set (if
-  // any) form a tree rooted on the "primary" RUNTIME_FUNCTION. We merge all such RUNTIME_FUNCTIONs
-  // from the same tree into a single UnwindRange.
+  uint64_t last_primary_address{};
+  uint64_t largest_primary_address{};
+  uint64_t num_chained_not_with_largest_primary = 0;
   for (const llvm::Win64EH::RuntimeFunction& runtime_function : runtime_functions) {
     auto primary_runtime_function_or_error = GetPrimaryRuntimeFunction(&runtime_function);
     if (primary_runtime_function_or_error.has_error()) {
-      return ErrorMessage{absl::StrFormat("Unable to load unwind info ranges: %s",
-                                          primary_runtime_function_or_error.error().message())};
+      return ErrorMessage{
+          absl::StrCat(kErrorMessagePrefix, primary_runtime_function_or_error.error().message())};
     }
     const llvm::Win64EH::RuntimeFunction* primary_runtime_function =
         primary_runtime_function_or_error.value();
 
+    const uint64_t start_address = object_file_->getImageBase() + runtime_function.StartAddress;
+    const uint64_t end_address = object_file_->getImageBase() + runtime_function.EndAddress;
     const uint64_t primary_address =
         object_file_->getImageBase() + primary_runtime_function->StartAddress;
-    const uint64_t end_address = object_file_->getImageBase() + runtime_function.EndAddress;
 
     if (unwind_ranges.empty()) {
       unwind_ranges.emplace_back(UnwindRange{primary_address, end_address});
+      ORBIT_CHECK(last_primary_address == 0);
+      ORBIT_CHECK(largest_primary_address == 0);
+      last_primary_address = primary_address;
+      largest_primary_address = primary_address;
       continue;
     }
 
+    ORBIT_CHECK(last_primary_address != 0);
+    ORBIT_CHECK(largest_primary_address != 0);
+
+    UnwindRange& last_range = unwind_ranges.back();
+    const uint64_t last_end_address = last_range.end;
+
+    if (end_address < start_address) {
```

Comment:
I don't find it obvious why this check is down here and not further up. Would it be wrong to check for this error case right after getting the start and end address (that is, right after line 457)? It would be different than the current code (unless there are some non-obvious implications that make the following impossible), as you may still create an unwind address range using `primary_address` and `end_address` right now, even if `end_address < start_address`. But that seems like you a using potentially bad data from an illegal runtime function?

Reply:
Yeah, and the same applied to the "chained RUNTIME_FUNCTION is not a previous one" case. I moved both up, should be better for the reason you mentioned.

Reply:
Ok, looks good